### PR TITLE
fix(cometd): Update Message interface for CometD 4.0

### PR DESCRIPTION
**Pull request by swese44**
- Adds BaseMessage interface to include required and optional properties common to all Messages.
- Adds SuccessfulHandshakeMessage, UnsuccessfulHandshakeMessage, HandshakeMessage and SubscribeMessage interfaces with their appropriate fields.
- Replaces Message interface with a union type of the other messager interfaces.
- Adds specific Listener interfaces and uses them on the appropriate handlers.
- Implements all fields documented at https://docs.cometd.org/current4/reference/#_bayeux_message_fields, and SuccesfulHandshakeMessage's undocumented `reestablish` field implemented here: https://github.com/cometd/cometd/blob/bb64aa2dee3e56c63fcbd46e3951abe278417851/cometd-javascript/common/src/main/webapp/js/cometd/cometd.js#L2180
- Adds some formatting via prettier

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://docs.cometd.org/current4/reference/#_bayeux_message_fields
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

File: types/cometd/index.d.ts [hunk 1]
```diff
@@ -81,12 +82,57 @@ export interface Configuration {
...
+    clientId: string;
+    authSuccessful?: true;
+    reestablish: boolean;
+}
+
+export interface UnsuccessfulHandshakeMessage extends BaseMessage {
+    successful: false;
+    error: string;
+    supportedConnectionTypes?: ConnectionType[];
+    reestablish?: undefined;
```
Comment (swese44) [inline types/cometd/index.d.ts]: this is a convenience so consumers can access `handshakeMessage.reestablish` without using a typeguard

File: types/cometd/index.d.ts [hunk 2]
```diff
@@ -81,12 +82,57 @@ export interface Configuration {
...
+    version?: string;
+    minimumVersion?: string;
+}
+
+export interface SuccessfulHandshakeMessage extends BaseMessage {
+    successful: true;
+    version: string;
+    supportedConnectionTypes: ConnectionType[];
+    clientId: string;
+    authSuccessful?: true;
```
Comment (alxHenry) [inline types/cometd/index.d.ts]: `authSuccessful` is optional for success messages?

File: types/cometd/index.d.ts [hunk 3]
```diff
@@ -81,12 +82,57 @@ export interface Configuration {
...
 
-export interface Message {
+export type ConnectionType = 'long-polling' | 'callback-polling' | 'iframe' | 'flash';
+export type ReconnectAdvice = 'retry' | 'handshake' | 'none';
+
+export interface BaseMessage {
     successful: boolean;
-    data: any;
+    channel: string;
+    id?: string;
```
Comment (alxHenry) [inline types/cometd/index.d.ts]: This is a lot of optional fields. None of these are guaranteed?

File: types/cometd/index.d.ts (hunk 2 above)
Comment (swese44) [reply types/cometd/index.d.ts]: yes, looks like a legacy feature that is still used by some implementations
https://docs.cometd.org/current4/reference/#_successful_handshake_response

File: types/cometd/index.d.ts (hunk 3 above)
Comment (swese44) [reply types/cometd/index.d.ts]: correct

**Comment by unindented** [comment]
Do we need to add anything to the test file to check these new type definitions and protect against regressions?

**Comment by swese44** [comment]
@unindented added

**Review by unindented — APPROVED** [review]

**Review by alxHenry — APPROVED** [review]

**Pull request merged**